File: apple/WordPressTV/Shared/Cards.swift
import SwiftUI

// The card vocabulary shared across Home, category grids, and search: a poster
// primitive plus the three card shapes in the design — landscape video card,
// portrait WordCamp event card, and the wide Continue Watching card with a
// resume bar.

/// A rounded poster image with a branded placeholder while it loads or if the
/// source has none. `aspect` sizes it; the caller fixes the width.
struct PosterImage: View {
    let url: URL?
    var aspect: CGFloat = 16 / 9
    var cornerRadius: CGFloat = 16

    var body: some View {
        AsyncImage(url: url) { phase in
            switch phase {
            case .success(let image):
                image.resizable().scaledToFill()
            default:
                ZStack {
                    Rectangle().fill(Color.white.opacity(0.08))
                    Image(systemName: "play.rectangle.fill")
                        .font(.largeTitle)
                        .foregroundStyle(.white.opacity(0.35))
                }
            }
        }
        .aspectRatio(aspect, contentMode: .fill)
        .frame(maxWidth: .infinity)
        .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
    }
}

/// Landscape poster + title. Used in every horizontal rail and grid. The poster
/// URL resolves lazily (private a8c.tv posters need a token appended), so only
/// on-screen cards fetch one. The `.card` button style gives the tvOS focus
/// lift/parallax for free.
struct VideoCard: View {
    let video: Video
    var width: CGFloat = 360
    let resolvePoster: (Video) async -> URL?
    let onSelect: () -> Void

    @State private var posterURL: URL?

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Button(action: onSelect) {
                PosterImage(url: posterURL)
            }
            .buttonStyle(.card)

            Text(video.title)
                .font(.callout.weight(.medium))
                .foregroundStyle(.white)
                .lineLimit(2, reservesSpace: true)
                .multilineTextAlignment(.leading)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .frame(width: width)
        .task { posterURL = await resolvePoster(video) }
    }
}

/// Portrait WordCamp event card. A real cover image from the event's latest
/// video sits behind a brand-tinted scrim, with the WordPress mark and event
/// name over it — a populated poster rather than a flat colour. Falls back to
/// the brand gradient until (or if) the cover resolves. Opens the event's videos.
struct PortraitCampCard: View {
    let event: ContentEvent
    var width: CGFloat = 270
    /// Resolves the cover image (the event's newest video poster). Lazy so only
    /// on-screen cards fetch.
    let resolveCover: (ContentEvent) async -> URL?
    let onSelect: () -> Void

    @State private var coverURL: URL?

    private var height: CGFloat { width * 1.5 }

    var body: some View {
        VStack(alignment: .leading, spacing: 18) {
            Button(action: onSelect) {
                ZStack(alignment: .bottomLeading) {
                    // Brand gradient — the base, and the fallback if no cover.
                    LinearGradient(colors: event.colors, startPoint: .top, endPoint: .bottom)

                    // Real cover image, sized to the card and clipped so it can't
                    // grow the ZStack (which would push the title out of view).
                    if let coverURL {
                        AsyncImage(url: coverURL) { image in
                            image.resizable().scaledToFill()
                        } placeholder: {
                            Color.clear
                        }
                        .frame(width: width, height: height)
                        .clipped()
                    }

                    // Full-card gradient: a light top darkening for cohesion,
                    // deepening into the brand colour at the bottom so the cover
                    // reads as this camp and the title stays legible over any art.
                    LinearGradient(
                        colors: [
                            .black.opacity(0.4),
                            .black.opacity(0.1),
                            event.colors.last?.opacity(0.98) ?? .black.opacity(0.9),
                        ],
                        startPoint: .top,
                        endPoint: .bottom
                    )

                    WordPressMark()
                        .frame(width: 36, height: 36)
                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
                        .padding(22)

                    VStack(alignment: .leading, spacing: 2) {
                        Text("WordCamp")
                            .font(.title3.weight(.semibold))
                            .foregroundStyle(.white.opacity(0.9))
                        Text(event.displayPlace)
                            .font(.system(size: 36, weight: .heavy))
                            .foregroundStyle(.white)
                            .minimumScaleFactor(0.6)
                            .lineLimit(2)
                    }
                    .padding(22)
                }
                .frame(width: width, height: height)
                .clipShape(RoundedRectangle(cornerRadius: 18))
            }
            .buttonStyle(.card)

            Text(event.name)
                .font(.callout.weight(.medium))
                .foregroundStyle(.white)
                .lineLimit(1)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .frame(width: width)
        .task { coverURL = await resolveCover(event) }
    }
}

/// Wide Continue Watching card: poster with a resume bar showing how far in the
/// viewer got. Tapping resumes from `progress.positionSeconds`.
struct ContinueWatchingCard: View {
    let progress: WatchProgress
    var width: CGFloat = 440
    let resolvePoster: @MainActor (WatchProgress) async -> URL?
    let onPosterResolved: @MainActor (String, URL) -> Void
    let onSelect: () -> Void

    @State private var posterURL: URL?

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Button(action: onSelect) {
                PosterImage(url: posterURL)
                    .overlay(alignment: .bottom) {
                        ResumeBar(fraction: progress.fractionComplete)
                            .padding(.horizontal, 12)
                            .padding(.bottom, 12)
                    }
            }
            .buttonStyle(.card)

            Text(progress.title)
                .font(.callout.weight(.medium))
                .foregroundStyle(.white)
                .lineLimit(2, reservesSpace: true)
                .multilineTextAlignment(.leading)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
        .frame(width: width)
        .task(id: "\(progress.videoGuid)-\(progress.posterURLString ?? "")") {
            posterURL = progress.posterURL
            if posterURL == nil, let resolved = await resolvePoster(progress) {
                posterURL = resolved
                onPosterResolved(progress.videoGuid, resolved)
            }
        }
    }
}

/// A thin rounded resume indicator: full-width track, brand-blue fill.
private struct ResumeBar: View {
    let fraction: Double

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(.black.opacity(0.5))
                Capsule()
                    .fill(Brand.blue)
                    .frame(width: max(6, geo.size.width * fraction))
            }
        }
        .frame(height: 6)
    }
}

/// The official WordPress logo mark, used across the nav bar and WordCamp cards.
/// Rendered from the bundled vector asset as a template so it tints to its
/// context and stays crisp at any size.
struct WordPressMark: View {
    var tint: Color = .white

    var body: some View {
        Image("WordPressLogo")
            .resizable()
            .renderingMode(.template)
            .aspectRatio(contentMode: .fit)
            .foregroundStyle(tint)
    }
}

/// Shared brand palette. The selected/accent blue is WordPress blue — chosen so
/// it stays distinct from the white tvOS focus highlight.
enum Brand {
    static let blue = Color(red: 0.22, green: 0.34, blue: 0.91)
}

File: android/app/src/main/kotlin/com/automattic/wordpresstv/ui/Cards.kt
package com.automattic.wordpresstv.ui

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.automattic.wordpresstv.R
import com.automattic.wordpresstv.catalog.wordCampCardColors
import com.automattic.wordpresstv.catalog.wordCampDisplayPlace
import com.automattic.wordpresstv.continuewatching.WatchProgress
import com.automattic.wordpresstv.core.domain.ContentEvent
import com.automattic.wordpresstv.core.domain.Video
import com.automattic.wordpresstv.ui.theme.BrandBlue
import androidx.tv.material3.Card
import androidx.tv.material3.CardDefaults
import androidx.tv.material3.Text

/**
 * The card vocabulary shared across Home, category grids, and search: a poster
 * primitive plus the three card shapes in the design — landscape video card,
 * portrait WordCamp event card, and the wide Continue Watching card with a
 * resume bar. Mirrors the Apple `Cards.swift`.
 */

private val PosterPlaceholder = Color(0xFF15151A)

/** A poster image over a branded placeholder, shown while it loads or if absent. */
@Composable
fun PosterBox(url: String?, modifier: Modifier = Modifier) {
    Box(modifier.background(PosterPlaceholder), contentAlignment = Alignment.Center) {
        if (url != null) {
            AsyncImage(
                model = url,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier.fillMaxSize(),
            )
        }
    }
}

/**
 * Landscape poster + title. Used in every horizontal rail and grid. The poster
 * URL resolves lazily (private a8c.tv posters need a token appended), so only
 * on-screen cards fetch one. The TV [Card] gives the focus lift for free.
 *
 * [modifier] sizes the card (a fixed width in rails, the cell width in grids);
 * [focusRequester], when set, lands initial focus on this card so the D-pad works.
 */
@Composable
fun VideoCard(
    video: Video,
    resolvePoster: suspend (Video) -> String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    focusRequester: FocusRequester? = null,
) {
    var posterUrl by remember(video.id) { mutableStateOf<String?>(null) }
    LaunchedEffect(video.id) { posterUrl = resolvePoster(video) }

    Column(modifier) {
        Card(onClick = onClick, modifier = Modifier.fillMaxWidth().focusRequesterOrNone(focusRequester)) {
            PosterBox(posterUrl, Modifier.fillMaxWidth().aspectRatio(16f / 9f))
        }
        Spacer(Modifier.height(10.dp))
        Text(
            text = video.title,
            color = Color.White,
            fontSize = 16.sp,
            fontWeight = FontWeight.Medium,
            minLines = 2,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

/**
 * Portrait WordCamp event card. A real cover image from the event's latest
 * video sits behind a brand-tinted scrim, with the WordPress mark and event name
 * over it — a populated poster rather than a flat colour. Falls back to the brand
 * gradient until (or if) the cover resolves. Opens the event's videos.
 */
@Composable
fun PortraitCampCard(
    event: ContentEvent,
    resolveCover: suspend (ContentEvent) -> String?,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    focusRequester: FocusRequester? = null,
) {
    var coverUrl by remember(event.slug) { mutableStateOf<String?>(null) }
    LaunchedEffect(event.slug) { coverUrl = resolveCover(event) }
    val colors = event.wordCampCardColors

    Column(modifier) {
        Card(
            onClick = onClick,
            modifier = Modifier.fillMaxWidth().height(230.dp).focusRequesterOrNone(focusRequester),
            scale = CardDefaults.scale(focusedScale = 1.03f),
        ) {
            Box(Modifier.fillMaxSize()) {
                // Brand gradient — the base, and the fallback if no cover.
                Box(Modifier.fillMaxSize().background(Brush.verticalGradient(colors)))

                // Real cover image, sized to the card and cropped.
                if (coverUrl != null) {
                    AsyncImage(
                        model = coverUrl,
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier.fillMaxSize(),
                    )
                }

                // Full-card scrim: a light top darkening for cohesion, deepening
                // into the brand colour at the bottom so the cover reads as this
                // camp and the title stays legible over any art.
                Box(
                    Modifier.fillMaxSize().background(
                        Brush.verticalGradient(
                            listOf(
                                Color.Black.copy(alpha = 0.4f),
                                Color.Black.copy(alpha = 0.1f),
                                colors.last().copy(alpha = 0.98f),
                            ),
                        ),
                    ),
                )

                WordPressMark(
                    modifier = Modifier.align(Alignment.TopStart).padding(18.dp).size(28.dp),
                )

                Column(Modifier.align(Alignment.BottomStart).padding(18.dp)) {
                    Text("WordCamp", color = Color.White.copy(alpha = 0.9f), fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
                    Text(event.wordCampDisplayPlace, color = Color.White, fontSize = 26.sp, fontWeight = FontWeight.ExtraBold, maxLines = 2, overflow = TextOverflow.Ellipsis)
                }
            }
        }
        Spacer(Modifier.height(8.dp))
        Text(event.name, color = Color.White, fontSize = 15.sp, fontWeight = FontWeight.Medium, maxLines = 1, overflow = TextOverflow.Ellipsis)
    }
}

/**
 * Wide Continue Watching card: poster with a resume bar showing how far in the
 * viewer got. Tapping resumes from [WatchProgress.positionMs].
 */
@Composable
fun ContinueWatchingCard(
    progress: WatchProgress,
    resolvePoster: suspend (WatchProgress) -> String?,
    onPosterResolved: (String, String) -> Unit,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    focusRequester: FocusRequester? = null,
) {
    var posterUrl by remember(progress.videoGuid) { mutableStateOf(progress.posterUrl) }
    LaunchedEffect(progress.videoGuid, progress.posterUrl) {
        posterUrl = progress.posterUrl
        if (posterUrl == null) {
            val resolved = resolvePoster(progress)
            if (resolved != null) {
                posterUrl = resolved
                onPosterResolved(progress.videoGuid, resolved)
            }
        }
    }

    Column(modifier) {
        Card(onClick = onClick, modifier = Modifier.width(320.dp).focusRequesterOrNone(focusRequester)) {
            Box(Modifier.fillMaxWidth().aspectRatio(16f / 9f)) {
                PosterBox(posterUrl, Modifier.fillMaxSize())
                ResumeBar(
                    fraction = progress.fractionComplete.toFloat(),
                    modifier = Modifier.align(Alignment.BottomCenter).fillMaxWidth().padding(12.dp),
                )
            }
        }
        Spacer(Modifier.height(10.dp))
        Text(
            text = progress.title,
            color = Color.White,
            fontSize = 16.sp,
            fontWeight = FontWeight.Medium,
            minLines = 2,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

/** A thin rounded resume indicator: full-width track, brand-blue fill. */
@Composable
private fun ResumeBar(fraction: Float, modifier: Modifier = Modifier) {
    Box(modifier.height(6.dp).clip(RoundedCornerShape(3.dp)).background(Color.Black.copy(alpha = 0.5f))) {
        Box(
            Modifier
                .fillMaxHeight()
                .fillMaxWidth(fraction.coerceIn(0f, 1f))
                .clip(RoundedCornerShape(3.dp))
                .background(BrandBlue),
        )
    }
}

/**
 * The official WordPress logo mark, used across the nav bar and WordCamp cards.
 * Rendered from the bundled vector so it tints to its context and stays crisp at
 * any size.
 */
@Composable
fun WordPressMark(modifier: Modifier = Modifier, tint: Color = Color.White) {
    Image(
        painter = painterResource(R.drawable.ic_wordpress_logo),
        contentDescription = null,
        colorFilter = ColorFilter.tint(tint),
        modifier = modifier,
    )
}

/** Apply [focusRequester] when present; otherwise leave the modifier untouched. */
private fun Modifier.focusRequesterOrNone(focusRequester: FocusRequester?): Modifier =
    if (focusRequester != null) this.focusRequester(focusRequester) else this
